### PR TITLE
Nodejs/week2

### PR DESCRIPTION
**Pull request by gebremedhin85**

### CODE REVIEW

File: nodejs/week2/mealSharing/.gitignore [hunk 1]
```diff
@@ -0,0 +1,8 @@
+
+node_modules
+
+src/bakckend/routes/cheap-meals.js
+src/bakckend/routes/large-meals.js
+src/bakckend/routes/meal.js
+src/bakckend/routes/reservation.js
```
Comment (Vintharas) [inline nodejs/week2/mealSharing/.gitignore:7]: hehe interesting. Why are these here? ^^

File: nodejs/week2/mealSharing/package.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+	},
+	"author": "",
+	"license": "ISC",
+	"devDependencies": {
+		"nodemon": "^2.0.4"
+	},
+	"dependencies": {
+		"express": "^4.17.1"
+	}
+}
```
Comment (Vintharas) [inline nodejs/week2/mealSharing/package.json:19]: Interesting! Are you using TABs for indentation instead of spaces? What operative system and editor are you using?

File: nodejs/week2/mealSharing/src/bakckend/index.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+const cheapMealsRoute = require("./routes/cheap-meals");
+const reservationsRoute = require("./routes/reservations");
+const singleReservationRoute = require("./routes/reservation");
+const reviewsRoute = require("./routes/reviews");
+
+const app = express();
+const port = 3000;
+
+app.use("/meals", mealsRoute);
+app.get("/meal", aMealRoute);
```
Comment (Vintharas) [inline nodejs/week2/mealSharing/src/bakckend/index.js:14]: this should be `/meals/:id`

File: nodejs/week2/mealSharing/src/bakckend/index.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+app.get("/large-meals", largeMealsRoute);
+app.get("/cheap-meals", cheapMealsRoute);
+app.use("/reservations", reservationsRoute);
+app.get("/reservation", singleReservationRoute);
```
Comment (Vintharas) [inline nodejs/week2/mealSharing/src/bakckend/index.js:18]: this should be `/reservations/:id`

File: nodejs/week2/mealSharing/src/bakckend/index.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+app.use("/reviews", reviewsRoute);
+
```
Comment (Vintharas) [inline nodejs/week2/mealSharing/src/bakckend/index.js:20]: there should be a `/reviews/:id` to get a single review by id

File: nodejs/week2/mealSharing/src/bakckend/index.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+	res.send(reservationsRoute);
+});
+
+app.get("/", (req, res) => {
+	res.send(singleReservationRoute);
+});
+
+app.get("/", (req, res) => {
+	res.send(reviewsRoute);
+});
```
Comment (Vintharas) [inline nodejs/week2/mealSharing/src/bakckend/index.js:47]: why these?

File: nodejs/week2/mealSharing/src/bakckend/routes/meals.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+router.get("/", (req, res) => {
+	const maxPrice = req.query.maxPrice;
+	const title = req.query.title;
+	const createdAfter = req.query.createdAfter;
+	const limit = req.query.limit;
+
+	if (maxPrice) {
+		const maxPriceValue = parseInt(maxPrice);
+		if (!maxPriceValue) {
+			res.status(404).send(`Enter price in number`);
```
Comment (Vintharas) [inline nodejs/week2/mealSharing/src/bakckend/routes/meals.js:14]: not sending a maxPriceValue is OK, this is an optional query parameter. If it is not sent with the request you just don't use it to filter the meals.

File: nodejs/week2/mealSharing/src/bakckend/routes/meals.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
+const express = require("express");
+const router = express.Router();
+const meals = require("../data/reviewed-meals");
+
+router.get("/", (req, res) => {
+	const maxPrice = req.query.maxPrice;
+	const title = req.query.title;
+	const createdAfter = req.query.createdAfter;
+	const limit = req.query.limit;
```
Comment (Vintharas) [inline nodejs/week2/mealSharing/src/bakckend/routes/meals.js:9]: in general, all of these query parameters are optional and could be provided or not with the request. If they are provided then you can use them to filter the meals according to `maxPrice`, `title`, etc. But if they are not provided you just don't need to use them. e.g.

`/meals` => give me all meals
`/meals?maxPrice=100` => give me all cheap meals under 100 crowns
`/meals?title=chicken` => give me all meals with chicken
`/meals?title=chicken&maxPrice=100` => give me all cheap meals with chicken

When the service behaves in this way in can be really useful because it provides a very rich functionality for the user.

File: nodejs/week2/mealSharing/src/bakckend/routes/meals.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+	if (maxPrice) {
+		const maxPriceValue = parseInt(maxPrice);
```
Comment (Vintharas) [inline nodejs/week2/mealSharing/src/bakckend/routes/meals.js:12]: you parsed the price value to a number. Nice! But you are still not checking whether it is a number or not. 

What happens if you try this? parseInt('hello world')

Comment (Vintharas) [reply nodejs/week2/mealSharing/src/bakckend/routes/meals.js:12]: or if a user sends a request to `/meals?maxPrice=hello`

File: nodejs/week2/mealSharing/src/bakckend/routes/meals.js [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+				res.json(mealsInRangePrice);
+			}
+		}
+	} else if (title) {
+		const formatedTitle = title.replace(/[.*%+?^${}()|[\]\\]/g, "\\$&");
+		const regularExpresion = new RegExp(
+			"(?:^|\\W)" + formatedTitle + "(?:$|\\W)",
+			"i"
+		);
+		const mealsWithTitle = meals.filter((m) => regularExpresion.test(m.title));
```
Comment (Vintharas) [inline nodejs/week2/mealSharing/src/bakckend/routes/meals.js:29]: you can substitute this for meals.filter(m => m.title.includes(title))

File: nodejs/week2/mealSharing/src/bakckend/routes/meals.js [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+		if (mealsWithTitle.length > 0) {
+			res.json(mealsWithTitle);
+		} else {
+			res.send(`There is no meal that has the word ${title} in its title`);
+		}
+	} else if (createdAfter) {
+		const date = Date.parse(createdAfter);
+		
+		if (!date && !(date instanceof Date)) {
+			res.status(404).send("Input is not of type date");
```
Comment (Vintharas) [inline nodejs/week2/mealSharing/src/bakckend/routes/meals.js:39]: ditto. as above

File: nodejs/week2/mealSharing/src/bakckend/routes/meals.js [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+			const mealsCreatedAfter = meals.filter(
+				(m) => Date.parse(m.createdAt) > date
+			);
+			if (mealsCreatedAfter.length > 0) res.send(mealsCreatedAfter);
+			else res.send(`There is no meal created after this date: ${date}`);
+		}
+	} else if (limit) {
+		const limitValue = parseInt(limit);
+		if (!limitValue) {
+			res.status(404).send(`Enter a limit in number`);
```
Comment (Vintharas) [inline nodejs/week2/mealSharing/src/bakckend/routes/meals.js:50]: ditto. as above

File: nodejs/week2/mealSharing/src/bakckend/routes/meals.js (hunk 8 above)
Comment (Vintharas) [reply nodejs/week2/mealSharing/src/bakckend/routes/meals.js:9]: I don't think if it was clear above but, in this function you're only handling one single query param:

e.g. if you provide `maxPrice` that's the only one that applies

You need to be able to combine all the query params and have them filter the meals.

File: nodejs/week2/mealSharing/src/bakckend/routes/reservations.js [hunk 13]
```diff
@@ -0,0 +1,25 @@
+const express = require("express");
+const router = express.Router();
+const reservations = require("../data/reservations.json");
+
+router.get("/:id", (req, res) => {
+	const reserveId = parseInt(req.params.id);
+	if (!reserveId) {
```
Comment (Vintharas) [inline nodejs/week2/mealSharing/src/bakckend/routes/reservations.js:7]: What happens if id = 0?

File: nodejs/week2/mealSharing/src/bakckend/routes/reservations.js [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+		res.send(`Enter an id number`);
+	} else {
+		const reservationWithThisId = reservations.find((r) => r.id === reserveId);
+		if (!reservationWithThisId) {
+			res.send(`There is no reservation with this id: ${reserveId}`);
```
Comment (Vintharas) [inline nodejs/week2/mealSharing/src/bakckend/routes/reservations.js:12]: In REST there's a formal way in which to tell a client that a resource with a given id doesn't exist.

Try to see how to do that with express: Send a 404 resource not found response

File: nodejs/week2/mealSharing/src/bakckend/routes/reservations.js [hunk 15]
```diff
@@ -0,0 +1,25 @@
+const express = require("express");
+const router = express.Router();
+const reservations = require("../data/reservations.json");
+
+router.get("/:id", (req, res) => {
+	const reserveId = parseInt(req.params.id);
+	if (!reserveId) {
+		res.send(`Enter an id number`);
```
Comment (Vintharas) [inline nodejs/week2/mealSharing/src/bakckend/routes/reservations.js:8]: In REST there's a formal way in which to tell a client that the parameters of the request are invalid.

Try to see how to do this with express: Send a 400 bad request error response

File: nodejs/week2/mealSharing/src/bakckend/routes/reviews.js [hunk 16]
```diff
@@ -0,0 +1,23 @@
+const express = require("express");
```
Comment (Vintharas) [inline nodejs/week2/mealSharing/src/bakckend/routes/reviews.js:1]: ditto. same as in previous file

File: nodejs/week2/warmup/index.js [hunk 17]
```diff
@@ -0,0 +1,22 @@
+const express = require("express");
+const app = express();
+
+app.get("/numbers/add", (req, res) => {
+	const firstNum = parseInt(req.query.first);
+	const secondNum = parseInt(req.query.second);
+
+	if (!firstNum) res.send(`Please enter first as a number`);
+	else if (!secondNum) res.send(`Please enter second as a number`);
+	else res.send(`${firstNum + secondNum}`);
```
Comment (Vintharas) [inline nodejs/week2/warmup/index.js:10]: what happens if firstNum or secondNum are 0?

File: nodejs/week2/warmup/index.js [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+});
+
+app.get("/numbers/multiply/:firstNum/:secondNum", (req, res) => {
+	const firstNum = parseInt(req.params.firstNum);
+	const secondNum = parseInt(req.params.secondNum);
+
+	if (!firstNum) res.send("Please enter a first number");
+	else if (!secondNum) res.send("Please enter a second number");
+	else res.send(`${firstNum * secondNum}`);
```
Comment (Vintharas) [inline nodejs/week2/warmup/index.js:19]: ditto

**Comment by gebremedhin85** [comment]
Thank you so much for the valuable feedback and your support. I will
implement them.
